src/contracts/core/DelegationManager.sol

```diff
@@ -528,131 +633,224 @@ contract DelegationManager is
-                "DelegationManager._completeQueuedWithdrawal: input length mismatch"
-            );
+            // complete the withdrawal by converting shares to tokens
+            _completeReceiveAsTokens(withdrawal, tokens, totalMagnitudes);
+        } else {
+            // Award shares back in StrategyManager/EigenPodManager.
+            _completeReceiveAsShares(withdrawal, tokens, totalMagnitudes);
         }
 
         // Remove `withdrawalRoot` from pending roots
         delete pendingWithdrawals[withdrawalRoot];
+        emit WithdrawalCompleted(withdrawalRoot);
+    }
 
-        if (receiveAsTokens) {
-            // Finalize action by converting shares to tokens for each strategy, or
-            // by re-awarding shares in each strategy.
-            for (uint256 i = 0; i < withdrawal.strategies.length;) {
-                require(
-                    withdrawal.startBlock + strategyWithdrawalDelayBlocks[withdrawal.strategies[i]] <= block.number,
-                    "DelegationManager._completeQueuedWithdrawal: withdrawalDelayBlocks period has not yet passed for this strategy"
-                );
+    /**
+     * @dev This function completes a queued withdrawal for a staker by converting shares to tokens and transferring to the withdrawer.
+     * This will apply any slashing that has occurred since the scaledShares were initially set in the Withdrawal struct
+     * by reading the original delegated operator's totalMagnitude at the time of withdrawal completion.
+     */
+    function _completeReceiveAsTokens(
+        Withdrawal calldata withdrawal,
+        IERC20[] calldata tokens,
+        uint64[] memory totalMagnitudes
+    ) internal {
+        // Finalize action by converting scaled shares to tokens for each strategy, or
+        // by re-awarding shares in each strategy.
+        for (uint256 i = 0; i < withdrawal.strategies.length;) {
+            require(
+                withdrawal.startTimestamp + strategyWithdrawalDelays[withdrawal.strategies[i]] <= block.timestamp,
+                "DelegationManager._completeReceiveAsTokens: withdrawalDelay period has not yet passed for this strategy"
+            );
 
-                _withdrawSharesAsTokens({
-                    staker: withdrawal.staker,
-                    withdrawer: msg.sender,
-                    strategy: withdrawal.strategies[i],
-                    shares: withdrawal.shares[i],
-                    token: tokens[i]
-                });
-                unchecked {
-                    ++i;
-                }
+            // Take already scaled staker shares and scale again according to current operator totalMagnitude
+            // This is because the totalMagnitude may have changed since withdrawal was queued and the staker shares
+            // are still susceptible to slashing
+            uint256 sharesToWithdraw =
+                _calculateSharesToCompleteWithdraw(withdrawal.scaledShares[i], totalMagnitudes[i]);
```

Comment:
library for this?

Reply:
https://github.com/Layr-Labs/eigenlayer-contracts/pull/679/commits/a70ed62f13fa8be58c456c8d518ec71e7a8482ab
Added back SlashingLib but moved `DEALOCATION_DELAY` and `MIN_WITHDRAWAL_DELAY` to immutables